src/Server.cpp

```diff
@@ -200,73 +313,352 @@ void Server::acceptConnection() const
 {
-	LOG_DEBUG << "Handling connection on fd " << clientSock;
+	LOG_DEBUG << "Handling connection: " << connection.getClient() << " for server: " << connection.getServer();
 	// Handle client data
-	char buffer[BUFFER_SIZE];
+	char buffer[s_bufferSize];
 	HTTPRequest request;
 
 	request.method = MethodCount;
 	request.httpStatus = StatusOK;
 	request.shallCloseConnection = false;
 
-	Location location = {};
-	location.path = "/";
-	location.root = "/workspaces/webserv";
-	location.index = "index.html";
-
-	ServerConfig serverConfig;
-	serverConfig.locations.push_back(location);
-
-	ConfigFile configFile;
-	configFile.serverConfigs.push_back(serverConfig);
-
-	const ssize_t bytesRead = read(clientSock, buffer, BUFFER_SIZE);
+	const ssize_t bytesRead = recv(connection.getClient().fd, buffer, s_bufferSize, 0);
 	if (bytesRead < 0) {
 		std::cerr << "error: read\n";
-		close(clientSock);
+		close(connection.getClient().fd);
 	} else if (bytesRead == 0) {
 		// Connection closed by client
-		close(clientSock);
+		close(connection.getClient().fd);
 	} else {
-		m_requestStrings[clientSock] += buffer;
-		if (checkForCompleteRequest(clientSock)) {
-			LOG_DEBUG << "Received complete request: " << '\n' << m_requestStrings[clientSock];
+		m_connectionBuffers[connection.getClient().fd] += buffer;
+		if (checkForCompleteRequest(connection.getClient().fd)) {
+			LOG_DEBUG << "Received complete request: " << '\n' << m_connectionBuffers[connection.getClient().fd];
 			try {
-				parser.parseHttpRequest(m_requestStrings[clientSock], request);
-				parser.clearParser();
+				m_requestParser.parseHttpRequest(m_connectionBuffers[connection.getClient().fd], request);
+				m_requestParser.clearParser();
 			} catch (std::exception& e) {
 				LOG_ERROR << "Error: " << e.what();
 			}
-			ResponseBuilder builder(configFile);
-			builder.buildResponse(request);
-			send(clientSock, builder.getResponse().c_str(), builder.getResponse().size(), 0 );
+			m_responseBuilder.buildResponse(request);
+			send(connection.getClient().fd, m_responseBuilder.getResponse().c_str(),
+				m_responseBuilder.getResponse().size(), 0);
+		} else {
+			LOG_DEBUG << "Received partial request: " << '\n' << m_connectionBuffers[connection.getClient().fd];
 		}
 	}
 }
 
+/**
+ * @brief Iterates through all connections and closes any that have timed out.
+
+ * The for loop through the connections map has no increment statement because the
+ * iterator is incremented in the loop body. If .erase() is called on an iterator,
+ * it is invalidated.
+ * The time since last event is saved in a variable to print it to the log and check
+ * if it is greater than m_clientTimeout. If it is, the connection is removed from epoll,
+ * and closed by calling Connection::closeConnection(). This makes sure that the connection
+ * is properly closed.
+ */
+void Server::handleTimeout()
+{
+	for (std::map<int, Connection>::iterator iter = m_connections.begin(); iter != m_connections.end();
+		/* no increment */) {
```

Comment:
This loop intentionally has the increment in the loop body. By calling std::map::erase() the passed iterator gets invalidated. This means we can't increment it anymore.
 
From cppreference: [References and iterators to the erased elements are invalidated. Other references and iterators are not affected.](https://en.cppreference.com/w/cpp/container/map/erase: )

I read that for standard associative-containers you increment in the body instead:
--> https://stackoverflow.com/a/8234813
--> https://en.cppreference.com/w/cpp/container/map/erase (again - see example)

If the condition is true we call `m_connections.erase(iter++);` 
With post-increment we can pass the old iterator and increment it, before it gets invalidated.

Reply:
As far as I understand it:

When an iterator gets erased this iterator is then invalid. Therefore you use a post increment when you call the erase function, to erase the "old" iterator and skip to the "new" iterator. This has to be done in the body of the for loop. 

But what I meant was just an implementation as a while loop:

```cpp
        std::map<int, Connection>::iterator iter = m_connections.begin();
	while (iter != m_connections.end()) {
		time_t timeSinceLastEvent = iter->second.getTimeSinceLastEvent();
		LOG_DEBUG << iter->second.getClient() << ": Time since last event: " << timeSinceLastEvent;
		if (timeSinceLastEvent > m_clientTimeout) {
			LOG_INFO << "Connection timeout: " << iter->second.getClient();
			removeEvent(m_epfd, iter->first);
			iter->second.closeConnection();
			m_connections.erase(iter++);
		} else
			++iter;	
	}
```
I think this works the same as the for loop. It is not a big deal, just thought that it looks nicer.

Reply:
cppreference also has the exact same style :o